Docker rewrite

jonseitz opened the pull request:
Continuing on seas-computing/course-planner#6.

This PR completely wipes out the Vagrant and Ansible setup in favor of a Docker-based development environment. This required changes to the the webpack configurations, which has been consolidated down into a single `webpackfile.js` that just handles the build process (the development server will be configured in a future PR). In the interest of further consolidating configurations, I've removed the mocha test-runner and replaced it with a `.mocharc.js` file (which probably should have been a separate PR. sorry!)

Note that you should be able to build the docker container successfully, but it will throw errors if you run it because the actual code in the `src` directory still needs to be updated (ditto on the tests)

jonseitz committed:
remove vagrant setup

jonseitz committed:
add Docker config

jonseitz committed:
Replace config dir with webpackfile and mocharc

jonseitz committed:
Create dummy .env file

Lays out basic environment variables that need to exist for docker
configs

jonseitz committed:
Update dependencies and scripts

Bumps all of the existing deps to latest major version. also
significantly simplies the scripts to build/test/start

jonseitz committed:
Tell nodemon to only watch server code

While using nodemon for serverside development, we don't need to
restart when the client code or tests change

jonseitz committed:
Remove mongo_init.js file mount

This initialization script shouldn't be necessary since we're moving
away from keeping a local user authorization database.

jonseitz committed:
Provide docker instructions

Removes existing configuration instructions; replaces them with
docker-compose up.

jonseitz committed:
Prune variables

The compose-specific values are unnecessary as we're using the default
`docker-compose.yml` location, and we no longer need a separate port for
the client or environment for babel.

jonseitz committed:
remove client port from compose file

jonseitz committed:
add esm for mocha

rmainwork commented on .mochainit.js:
```diff
@@ -0,0 +1,16 @@
+import Enzyme from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+
+before(function configureEnzyme() {
```
Is there any reason not to use arrow functions here?

jonseitz replied:
Not really, since we're not using `this`, but the `function` syntax is more consistent with our other mocha files.

rmainwork commented on webpackfile.js:
```diff
@@ -0,0 +1,56 @@
+const path = require('path');
```
Maybe de-structure this?

```suggestion
const { resolve } = require('path');
```

jonseitz replied:
Yeah, that's a good point. Updated in 4297187

rmainwork commented on webpackfile.js:
```diff
@@ -0,0 +1,56 @@
+  name: 'client',
+  entry: ['./src/client/index.js'],
+  output: {
+    path: path.resolve(__dirname, 'build/static'),
```
```suggestion
    path: resolve(__dirname, 'build/static'),
```

jonseitz replied:
see 4297187

rmainwork commented on webpackfile.js:
```diff
@@ -0,0 +1,56 @@
+    './src/server/index.js',
+  ],
+  output: {
+    path: path.resolve(__dirname, 'build'),
```
```suggestion
    path: resolve(__dirname, 'build'),
```

jonseitz replied:
see 4297187

jonseitz committed:
Lock mongo to version 3.6

Per discussions about moving to DocumentDB, we should use mongo3.6
locally to maintain compatibility (at least until aws make documentDB
available as its own docker container).

rmainwork reviewed (changes requested):
Overall looks good. Couple of small things toward the end in the webpack config file. Also interested in the naming of `webpackfile.js`. Is there a reason for it being called that rather than the usual `webpack.config.js`?

jonseitz committed:
destructure path.resolve

jonseitz commented:
No particular reason for the change to `webpackfile.js` from `webpack.config.js`, but they're both read automatically by the webpack-cli and I have a slight preference for the shorter form without the double dot syntax. I'd be open to changing to the more common form for the sake of consistency if you think that makes more sense. (If you like to watch people argue there's [an old github issue][issue] about this).

I also made one additional change to `docker-compose.yml`, setting the mongo version to 3.6 for documentDB compatibility.

[issue]: https://github.com/webpack/webpack/issues/967

rmainwork commented:
Fair enough, I didn't know that webpack could find `webpackfile.js`, so I was wondering if you were somehow using some other tool to compile it down to `webpack.conf.js` or something. I did see the other issue in my research before I commented about the above :laughing:. This looks good to me. Merge at will.

jonseitz committed:
Merge branch 'master' into docker-rewrite
